README.md
# Checklist Rating
A Flutter package that provides a customizable checklist rating widget.

## Installation
Add checklist_rating to your pubspec.yaml dependencies:

```yaml
dependencies:
  checklist_rating: ^0.0.7
```

Then run:
```sh
flutter pub get
```

## Usage
To use the ChecklistSetRating widget, you first need to import the package:
```dart
import 'package:checklist_rating/checklist_rating.dart';
```

Here's an example of how you can integrate ChecklistSetRating into your Flutter app:
```dart
void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int rating = 0;
  bool isChecked = false;

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Checklist Rating Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: Scaffold(
        body: Center(
          child: Padding(
            padding: const EdgeInsets.all(20.0),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                ChecklistRating.custom(
                  titleChekbox: "Sample Checklist",
                  onCheckboxChanged: (value) {
                    // Handle set checklist here
                    setState(() {
                      isChecked = value ?? false;
                    });
                  },
                  setRating: rating,
                  setChecked: isChecked,
                  maxRatings: 5,
                  onRatingChecked: (rating) {
                    // Handle rating changes here
                    setState(() {
                      this.rating = rating;
                    });
                  },
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

```

## Example how to use checklist_rating
To use the ChecklistSetRating widget, you first need to import the package:

https://github.com/nunutech40/checklist_rating/assets/49187517/3ae6730f-13db-4f70-bc9c-70b9abecbb3d




CHANGELOG.md
## 0.0.1

- first realese, u can use checlist rating.

## 0.0.2

- add parameter to set checbok rating needed, 3 - 5 checkbox
- add readme for installation + usage guide

## 0.0.3

- refactor code for publish to pub dev - warning in 
```sh
flutter pub publish --dry-run
```

## 0.0.4

- change version for update readme installation

## 0.0.5
- refactor code without static, and add factory to class ChecklistSetrating to create different variant of ChecklistSetrating

## 0.0.6
- refactor checklistrating, there's error not implement oncheck + onsetrating

## 0.0.7
- remove redundant widget i forgot to delete.

test/checklist_rating_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:checklist_rating/checklist_rating.dart';

void main() {}


lib/widget/small_rating.dart
import 'package:flutter/material.dart';

import 'common/styles.dart';

class SmallRating extends StatelessWidget {
  final String rating;
  final bool isSelected;
  final VoidCallback? onTap; // Keep the same callback type

  const SmallRating(
      {Key? key,
      required this.rating,
      this.isSelected = false,
      this.onTap}) // Keep the same argument type
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap, // This stays the same
      child: Container(
        height: 24.0,
        width: 24.0,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: isSelected ? lighPrimaryColor : Colors.white,
          borderRadius: BorderRadius.circular(6.0),
          border: Border.all(
            color: isSelected ? primaryColor : e2Gray,
            width: 1.0,
          ),
        ),
        child: Text(
          rating,
          textAlign: TextAlign.center,
          style: const TextStyle(fontSize: 12)
              .copyWith(color: isSelected ? primaryColor : Colors.black),
        ),
      ),
    );
  }
}


lib/widget/checkbox_default.dart
import 'package:flutter/material.dart';

import 'common/styles.dart';

class CheckboxDeft extends StatefulWidget {
  final bool isChecked;
  final void Function(bool?)? onChanged;

  const CheckboxDeft({Key? key, this.isChecked = false, this.onChanged})
      : super(key: key);

  @override
  CheckboxDeftState createState() => CheckboxDeftState();
}

// You're not supposed to expose this class outside this file.
// If you need to use it outside this file, remove the underscore.
class CheckboxDeftState extends State<CheckboxDeft> {
  @override
  Widget build(BuildContext context) {
    Color getColor(Set<MaterialState> states) {
      const Set<MaterialState> interactiveStates = <MaterialState>{
        MaterialState.pressed,
        MaterialState.hovered,
        MaterialState.focused,
      };
      if (states.any(interactiveStates.contains)) {
        return secondaryColor;
      }
      return primaryColor;
    }

    return Checkbox(
      checkColor: Colors.white,
      fillColor: MaterialStateProperty.resolveWith(getColor),
      value: widget.isChecked,
      onChanged: widget.onChanged,
    );
  }
}


lib/widget/small_rating_white.dart
import 'package:flutter/material.dart';

import 'common/styles.dart';

class SmallRatingWhite extends StatelessWidget {
  final String rating;
  final bool isSelected;
  final VoidCallback? onTap; // Keep the same callback type

  const SmallRatingWhite(
      {Key? key,
      required this.rating,
      this.isSelected = false,
      this.onTap}) // Keep the same argument type
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap, // This stays the same
      child: Container(
        height: 24.0,
        width: 24.0,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: isSelected ? lighPrimaryColor : Colors.white,
          borderRadius: BorderRadius.circular(6.0),
          border: Border.all(
            color: isSelected ? primaryColor : e2Gray,
            width: 1.0,
          ),
        ),
        child: Text(
          rating,
          textAlign: TextAlign.center,
          style: const TextStyle(fontSize: 12)
              .copyWith(color: isSelected ? primaryColor : Colors.black),
        ),
      ),
    );
  }
}


lib/widget/item_row_set_setting.dart
import 'package:flutter/material.dart';

import 'checkbox_default.dart';
import 'common/extension.dart';
import 'small_rating_white.dart';

class ItemRowSetRating extends StatelessWidget {
  /// The title for the checkbox.
  final String? titleChekbox;

  /// The initial state of the checkbox.
  final bool setChecked;

  /// A callback when the checkbox state changes.
  final void Function(bool?)? onCheckboxChanged;

  /// A callback when a rating is selected.
  final void Function(int)? onRatingChecked;

  /// The initial rating value.
  final int setRating;

  /// The maximum number of ratings to show. Must be between 3 and 5.
  final int maxRatings;

  ItemRowSetRating({
    Key? key,
    this.titleChekbox,
    this.setChecked = false,
    this.onRatingChecked,
    this.setRating = 0,
    this.onCheckboxChanged,
    this.maxRatings = 5,
  }) : super(key: key) {
    if (maxRatings < 3 || maxRatings > 5) {
      throw ArgumentError(
          'maxRatings should be between 3 and 5 inclusive. Received: $maxRatings.');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 8.0),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              CheckboxDeft(
                isChecked: setChecked,
                onChanged: onCheckboxChanged,
              ),
              Text(
                truncateText(titleChekbox ?? 'Title Checkbox', 18),
                overflow: TextOverflow.ellipsis,
                style: const TextStyle(fontSize: 14),
              ),
            ],
          ),
        ),
        Padding(
          padding: const EdgeInsets.only(right: 24.0),
          child: Row(
            children: List<Widget>.generate(
              maxRatings,
              (index) => Row(
                children: [
                  SmallRatingWhite(
                    rating: '${index + 1}',
                    isSelected: setRating >= index + 1,
                    onTap: () => onRatingChecked?.call(index + 1),
                  ),
                  if (index != maxRatings - 1) const SizedBox(width: 8.0),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }
}


lib/checklist_rating.dart
library checklist_rating;

import 'package:checklist_rating/widget/checkbox_default.dart';
import 'package:checklist_rating/widget/common/extension.dart';
import 'package:checklist_rating/widget/small_rating.dart';
import 'package:flutter/material.dart';

/// `ChecklistRating` is a widget that presents a flexible row layout containing
/// a checkbox with its associated title on the left, and a series of rating widgets on the right.
///
/// This widget is tailored for scenarios where a user needs to both check an item and provide
/// a rating for it.

class ChecklistRating extends StatelessWidget {
  /// [titleChekbox] defines the text displayed next to the checkbox.
  /// If not provided, it will default to 'Title Checkbox' after truncation.
  final String? titleChekbox;

  /// [setChecked] indicates the initial checked state of the checkbox.
  /// Defaults to `false` if not explicitly set.
  final bool setChecked;

  /// [onCheckboxChanged] is a callback triggered when the checkbox state changes.
  /// It provides the new checkbox state.
  final void Function(bool?)? onCheckboxChanged;

  /// [onRatingChecked] is a callback triggered when a user selects a rating value.
  /// It returns the selected rating.
  final void Function(int)? onRatingChecked;

  /// [setRating] sets the initial selected rating value, where `0` implies no initial rating.
  final int setRating;

  /// [maxRatings] determines the maximum number of rating items to display.
  /// This value must range between 3 and 5, with the default being 5.
  final int maxRatings;

  ChecklistRating({
    Key? key,
    this.titleChekbox,
    this.setChecked = false,
    this.onRatingChecked,
    this.setRating = 0,
    this.onCheckboxChanged,
    this.maxRatings = 5,
  }) : super(key: key) {
    if (maxRatings < 3 || maxRatings > 5) {
      throw ArgumentError(
          'maxRatings should be between 3 and 5 inclusive. Received: $maxRatings.');
    }
  }

  // Factory constructor to create a custom ChecklistRating.
  factory ChecklistRating.custom({
    String? titleChekbox,
    int? setRating,
    bool? setChecked,
    int? maxRatings,
    void Function(bool?)? onCheckboxChanged,
    void Function(int)? onRatingChecked,
  }) {
    return ChecklistRating(
      titleChekbox: titleChekbox,
      setChecked: setChecked ?? false,
      setRating: setRating ?? 0,
      maxRatings: maxRatings ?? 5,
      onCheckboxChanged: onCheckboxChanged,
      onRatingChecked: onRatingChecked,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 8.0),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              CheckboxDeft(
                isChecked: setChecked,
                onChanged: onCheckboxChanged,
              ),
              Text(
                truncateText(titleChekbox ?? 'Title Checkbox', 18),
                overflow: TextOverflow.ellipsis,
                style: const TextStyle(fontSize: 14),
              ),
            ],
          ),
        ),
        Padding(
          padding: const EdgeInsets.only(right: 24.0),
          child: Row(
            children: List<Widget>.generate(
              maxRatings,
              (index) => Row(
                children: [
                  SmallRating(
                    rating: '${index + 1}',
                    isSelected: setRating >= index + 1,
                    onTap: () => onRatingChecked?.call(index + 1),
                  ),
                  if (index != maxRatings - 1) const SizedBox(width: 8.0),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }
}
